server/events.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from enum import Enum
+from typing import Literal, TypedDict
+
+from pydantic import BaseModel, validator
+
+from .codes import StatusCode
+
+Position = tuple[int, int]
+Replacement = TypedDict("Replacement", {"from": int, "to": int, "value": str})
+
+
+class EventType(str, Enum):
+    """The type of a WebSocket event.
+
+    It is declared as a subclass of str to help serialization.
+    """
+
+    CONNECT = "connect"
+    DISCONNECT = "disconnect"
+    SYNC = "sync"
+    MOVE = "move"
+    REPLACE = "replace"
+    ERROR = "error"
+
+
+class EventData(BaseModel):
+    """The data of a WebSocket event.
+
+    This is just a base class that other classes will inherit from.
+    """
+
+
+class ConnectData(EventData):
+    """The data of a connection event.
+
+    Fields:
+        connection_type: "create" if the user wants to create the room, "join"
+            if the user wants to join the room.
+        difficulty (optional): The difficulty of the room, only needed if the
+            "connection_type" is "create".
+        room_code: The unique four-letters code that will represent the room.
+        username: The username of the user creating or joining the room.
+    """
+
+    connection_type: Literal["create", "join"]
+    difficulty: int | None
```

Comment:
shouldn't difficulty be an `Optional` type rather then requiring it but possible as None?

Reply:
> shouldn't difficulty be an `Optional` type rather then requiring it but possible as None?

`Optional` is essentially an alias for `Union[thing, None]`. In this case it would be `Union[int, None]` which in Python 3.10 syntax is `int | None`

Reply:
> shouldn't difficulty be an `Optional` type rather then requiring it but possible as None?

`Optional` is equal to `Type | None`. `pydantic` is smart enough to know that optional fields are not required.

If you try to send a connection event without `difficulty` it works fine.